### ISSUE TITLE
Conversion to a different "known" contract type is possible

### ISSUE THREAD

**Issue by androlo** [ethereum/solidity#4308]
Using Remix, `version:0.4.24+commit.e67f0147.Emscripten.clang`.

This code will compile.

```
contract A {
    function f() public returns (uint) {
    }
}

contract B {
    function g() public returns (bool) {
    }
}

contract C {
    function h() public returns (bool) {
        B b = B(new A());
        b.g();
    }
}
```
Guessing that the issue is that it will implicitly cast `new A()` to an address. Not casting `A` will cause a compiler error, i.e. `B b = new A()`.

There is not much information about the contract type in the docs (though I may have missed something) but given that the code below fails, it seems like typing is nominal:

```
contract A {
}

contract B {
}

contract C {
    function h() public returns (bool) {
        B b = new A();
    }
}
```

Regardless, it seems like the code in the first example should fail.

**Comment by chriseth** [comment]
Explicit conversions between all contract types is allowed, we might consider removing that (and forcing users to at least go through `address`).

**Comment by chriseth** [comment]
Will be fixed with 0.5.0 and tracked in #4320.
